### PR TITLE
bpo-42128: Add __match_args__ to structseq-based classes

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue42128

### CODE REVIEW

**Review by brandtbucher — APPROVED** [review]
Thanks Pablo!

Should we have a test for a structseq with unnamed fields? Asserting that `os.stat_result.__match_args__` is expected should do it.

File: Objects/structseq.c [hunk 1]
```diff
@@ -399,6 +400,31 @@ initialize_structseq_dict(PyStructSequence_Desc *desc, PyObject* dict,
...
+
+    // Prepare and set __match_args__
+    PyObject* keys = PyTuple_New(n_members - n_unnamed_members);
+    if (keys == NULL) {
+        return -1;
+    }
+
+    Py_ssize_t i, k;
+    for (i = k = 0; i < n_members; ++i) {
+        if (desc->fields[i].name == PyStructSequence_UnnamedField) {
```
Comment (pablogsal) [inline Objects/structseq.c:412]: I don't think we can handle this in a better way: just exclude them from __match_args__. The other (much worse option IMHO) is generate some missing names for these.

**Comment by pablogsal** [comment]
> Should we have a test for a structseq with unnamed fields? Asserting that `os.stat_result.__match_args__` is expected should do it.

Done in 40bf3b7

**Comment by pablogsal** [comment]
> Looking a bit closer... should we limit this to only `VISIBLE_SIZE_TP(type)` members? Since they're not displayed in the repr and can't be iterated over, I don't think there's any way to easily (as a user reading the code) map positional arguments to members like `tm_zone` and `tm_gmtoff`.
> 
> Plus, we can always extend it later if it turns out to be an unnecessary limitation.

Good point! I agree with that.  I have pushed a new commit with this limitation.

**Review by brandtbucher — APPROVED** [review]
Looks good to me!

The only improvement I might suggest is building up a list, then converting it to a tuple (or creating a tuple and shrinking it later with `_PyTuple_Resize`). That way we only need to loop once.

**Comment by pablogsal** [comment]
> Looks good to me!
> 
> The only improvement I might suggest is building up a list, then converting it to a tuple (or creating a tuple and shrinking it later with `_PyTuple_Resize`). That way we only need to loop once.

I considered that, but I think the code would be harder to read because appending to the list can fail on every iteration and we also need to covert the list to a tuple that can also fail. Also it will be less efficient so I decided with the current approach.

**Comment by pablogsal** [comment]
> (or creating a tuple and shrinking it later with `_PyTuple_Resize`). That way we only need to loop once.

Done

**Review by brandtbucher — APPROVED** [review]

**Pull request merged**